src/source-map.ts

```diff
@@ -0,0 +1,216 @@
+            (mapping.generatedLine === 1 ? firstLineCharOffset : 0)
+      }
+    };
+
+    if (mapping.name) {
+      newMapping.name = mapping.name;
+    }
+
+    generator.addMapping(newMapping);
+  });
+
+  return generator.toJSON();
+}
+
+export async function getExistingSourcemap(
+    analyzer: AnalysisContext|Analyzer,
+    sourceUrl: string,
+    sourceContent: string) {
+  const sourceMappingUrlParts = sourceContent.match(sourceMappingUrlExpr);
+  if (sourceMappingUrlParts === null) {
+    return null;
+  }
+
+  let sourcemap: RawSourceMap;
+  let mapUrl = sourceUrl;
+  const inlineSourcemapParts =
+      sourceMappingUrlParts[1].match(inlineSourceMapExpr);
+  if (inlineSourcemapParts !== null) {
+    sourcemap = base64StringToRawSourceMap(inlineSourcemapParts[2]);
+  } else {
+    mapUrl = urlLib.resolve(sourceUrl, sourceMappingUrlParts[1].trim());
+    sourcemap = JSON.parse(await analyzer.load(mapUrl)) as RawSourceMap;
+  }
+
+  // Rewrite the sources array to be relative to the current URL
+  if (sourcemap.sources) {
+    sourcemap.sources =
+        sourcemap.sources.map(source => urlLib.resolve(mapUrl, source));
+  }
+  return sourcemap;
+}
+
+/**
+ * For an inline script AST node, locate an existing source map url comment.
+ * If found, load that source map. If no source map url comment is found,
+ * create an identity source map.
+ *
+ * In both cases, the generated mappings reflect the relative position of
+ * a token within the script tag itself (rather than the document). This
+ * is because the final position within the document is not yet known. These
+ * relative positions will be updated later to reflect the absolute position
+ * within the bundled document.
+ */
+export async function addOrUpdateSourcemapComment(
+    analyzer: AnalysisContext|Analyzer,
+    sourceUrl: string,
+    sourceContent: string,
+    lineOffset: number,
+    firstLineCharOffset: number) {
+  let sourcemap =
+      await getExistingSourcemap(analyzer, sourceUrl, sourceContent);
+
+  let hasExisting = true;
+  if (sourcemap === null) {
+    hasExisting = false;
+    sourcemap = createJsIdentitySourcemap(
+        sourceUrl, sourceContent, lineOffset, firstLineCharOffset);
+  }
+
+  if (sourcemap === null) {
+    return sourceContent;
+  }
+
+  let updatedSourcemapComment =
+      inlineSourcemapPrefix + rawSourceMapToBase64String(sourcemap) + '\n';
+  if (hasExisting) {
+    return sourceContent.replace(sourceMappingUrlExpr, updatedSourcemapComment);
+  } else {
+    if (sourceContent.length > 0 &&
+        sourceContent[sourceContent.length - 1] === '\n') {
+      updatedSourcemapComment = updatedSourcemapComment.substr(1);
+    }
+
+    return sourceContent + updatedSourcemapComment;
+  }
+}
+
+
+/**
+ * Update mappings in source maps within inline script elements to reflect
+ * their absolute position within a bundle. Assumes existing mappings
+ * are relative to their position within the script tag itself.
```

Comment:
Is this assumption just due to where the source mapping takes place in the pipeline? i.e. that by the time it sees the source flowing through it's already been inlined so there's no way to distinguish external code that was inlined without updating its sourcemap from code that was e.g. in the original HTML and had its sourcemap?

Hm, and there's also the case of inlining some HTML which itself has inline JS which has a source map that has already been corrected for its place in that HTML file.

Asking just for clarity. Sourcemaps in inline javascript are rare enough, and this PR is already a significant enough chunk of work that we definitely shouldn't block on addressing it.